Add PolyhedralComplex as new object

lkastner opened the pull request:
Since we have cones and fans (and polyhedra), it makes sense to complete this.

Polyhedral complexes are needed in the tropical geometry part to come, nevertheless it seems sensible to have a separate PR.

lkastner committed:
PolyhedralGeometry: Start implementing PolyhedralComplex

lkastner committed:
PolyhedralGeometry: Start documentation of polyhedral complexes

lkastner commented:
Looking at https://github.com/oscar-system/Oscar.jl/actions/runs/1558072926 it talks about merging? Was it trying to merge my draft?

thofma commented:
Yes, it always runs the tests on a merge into the master branch. The failure was just a network hiccup.

lkastner committed:
PolyhedralGeometry: Some leftover direct access to underlying polymake object eliminated

lkastner committed:
PolyhedralGeometry: Access to basic properties of PolyhedralComplex

lkastner committed:
PolyhedralGeometry: Add some docstrings

lkastner committed:
PolyhedralGeometry: Various properties of PolyhedralComplex

lkastner committed:
PolyhedralGeometry: Add polyhedra_of_dim for extracting polyhedra from a polyhedral complex

lkastner committed:
PolyhedralGeometry: Add several functions for PolyhedralComplex

- Fix some renaming which was forgotten in exports
- Add new functions to docs
- nrays, nvertices, npolyhedra, f_vector added

lkastner committed:
PolyhedralGeometry: Add codim(PolyhedralComplex)

lkastner committed:
PolyhedralGeometry: isembedded(PolyhedralComplex) added

lkastner committed:
PolyhedralGeometry: Add some more basic functionality for PolyhedralComplex

Commit:
add intersection and k-skeleton of polyhedral complexes

lkastner committed:
PolyhedralGeometry: Fix bug in doctest

lkastner committed:
PolyhedralGeometry: Add visualization of PolyhedralComplex

lkastner committed:
PolyhedralGeometry: Serialize PolyhedralComplex

Commit:
Extended PolyhedralComplex constructor in order to make PolyhedralComplex(maximal_polyhedra(PC)) work properly

fingolfin commented on docs/src/PolyhedralGeometry/polyhedral_complexes.md:
```diff
@@ -0,0 +1,54 @@
+$\mathbb{F}^n$, for $n$ fixed, is a *polyhedral complex* if
+
+- the set $\mathcal{F}$ is closed with respect to taking faces and
+- if $C,D\in\mathcal{F}$ then $C\cap D$ is a face of both, $C$ and $D$.
```
```suggestion
- if $C,D\in\mathcal{F}$ then $C\cap D$ is a face of both $C$ and $D$.
```
Seems this PR was extracted from PR #889 so I guess some more of the comments I made there apply here (and should probably be addressed here).

Alas, this PR here *also* is marked as a draft. So I am not sure whether you actually want feedback right now, or not, or what?!

lkastner replied:
I am unsure why I am being questioned in this way. I marked both this and #889 as drafts since they are not finished. You are of course free to give feedback, but you might remark on things that we intend to change anyway. I thought marking a PR as _draft_ implies that feedback at this point is not needed yet.

fingolfin replied:
Hi Lars, sorry, only saw your comment here just now. To be clear: I did not mean to "question" you (like an inquisitor), to the contrary: what I meant was this: I saw the one PR and had some immediate feedback. Then I saw this PR. Then I noticed that they are marked as draft, and now was like "oh, ok, maybe he doesn't even want feedback?!?".

My apologies for very badly communicating this!

micjoswig commented on src/PolyhedralGeometry/PolyhedralComplex/constructors.jl:
```diff
@@ -0,0 +1,87 @@
+###############################################################################
+function Base.show(io::IO, PC::PolyhedralComplex)
+    ad = ambient_dim(PC)
+    if ad == -1.0
```
I don't understand this line of code: (1) it should not be possible to create a polyhedral complex without an ambient dimension. (2) Why is the right hand side `-1.0` (and not an integer)?

lkastner replied:
Could we not make a polyhedral complex just from a Hasse diagram? @YueRen says he needs polyhedral complexes that aren't embedded. We also had polyhedra without embedding. This right hand side stems from the fact that polymake would not error, but instead return `-1.0` if the ambient dimension could not be computed. The code is the same for the polyhedron, maybe this is fixed now, @benlorenz ?

benlorenz replied:
I fixed the code in polymake for Cone, Polytope and PolyhedralFan but somehow missed PolyhedralComplex. I will add the missing input property for the upcoming release, after that change the ambient dim method should fail instead of returning a plain perl number. The code tries to compute `FAN_AMBIENT_DIM - 1` which should be an int but in this case it is `undef - 1` which perl evaluates to a float (perl number type) -1. (This is then translated to a julia Float)

micjoswig commented on src/PolyhedralGeometry/PolyhedralComplex/standard_constructions.jl:
```diff
@@ -0,0 +1,91 @@
+function polyhedral_complex_workaround(pm::Polymake.BigObject)
+    pc = pm
+    typename = Polymake.type_name(pm)
+    if typename[1:13] == "PolyhedralFan"
```
This should be refactored throughout: test for an exact match (not a substring).

lkastner replied:
It is a workaround that goes away with the next polymake release, once 4.6 is in Julia.

micjoswig commented on src/PolyhedralGeometry/Serialization.jl:
```diff
@@ -141,3 +141,30 @@ function load_subdivisionofpoints(filename::String)
+function load_polyhedralcomplex(filename::String)
+   bigobject = Polymake.load_bigobject(filename)
+   typename = Polymake.type_name(bigobject)
+   if typename[1:17] != "PolyhedralComplex"
```
Same as above.

lkastner replied:
At some point the idea was to support more than just `Rational` of polymake in OSCAR. This is why I use the substring.

micjoswig reviewed (commented):
Generally looks good to me (except for minor complaints).  I'd suggest to fix those, merge from master and then to merge everything back into master.

micjoswig commented on docs/src/PolyhedralGeometry/polyhedral_complexes.md:
```diff
@@ -42,6 +42,7 @@ ambient_dim(PC::PolyhedralComplex)
 codim(PC::PolyhedralComplex)
 dim(PC::PolyhedralComplex)
 f_vector(PC::PolyhedralComplex)
+isembedded(PC::PolyhedralComplex)
```
Is there a specific reason for this function?

According to your definition in the doc a polyhedral complex is *always* embedded.

lkastner replied:
@YueRen needs complexes that aren't embedded, where we e.g. only know the Hasse diagram. I was going to say that we also have polyhedra that aren't embedded, but they seem to be gone? The definition seems now always to be embedded.

lkastner committed:
PolyhedralGeometry: Scalars are always Rational

lkastner committed:
PolyhedralGeometry: Workaround obsolete since new polymake release

lkastner committed:
PolyhedralGeometry: Adapt to improved conversion

lkastner committed:
PolyhedralGeometry: Add back export removed in botched rebase

lkastner committed:
PolyhedralGeometry: Incorporate first batch of comments by @fingolfin in #889

lkastner committed:
PolyhedralGeometry: Adapt show methods

Previously Polymake.jl did not give an error when the ambient dimension
could not be computed, instead it returned -1.0. This has now changed,
so we can deal with this properly in the show method.

lkastner committed:
PolyhedralGeometry: Incorporate second batch of comments by @fingolfin in #889

lkastner committed:
PolyhedralComplex: Add rays to docs

benlorenz commented on src/PolyhedralGeometry/PolyhedralComplex/properties.jl:
```diff
@@ -0,0 +1,431 @@
+julia> PC = PolyhedralComplex(IM, VR, far_vertices);
+
+julia> nvertices(PC)
+1
```
```suggestion
julia> nrays(PC)
3
```
example swapped with `nvertices`

benlorenz commented on src/PolyhedralGeometry/PolyhedralComplex/properties.jl:
```diff
@@ -0,0 +1,431 @@
+julia> PC = PolyhedralComplex(IM, VR, far_vertices);
+
+julia> nrays(PC)
+3
```
```suggestion
julia> nvertices(PC)
1
```

lkastner committed:
PolyhedralGeometry: Fix nrays and nvertices doc examples
